File: packages/canonical-json-ts/src/validate.ts
import { CanonicalJsonError } from './errors.js';
import { joinPath } from './path.js';
import type { FieldKind, FieldSchema, ObjectSchema } from './schema.js';

const HEX_RE = /^[0-9a-f]+$/;
const ADDRESS_RE = /^0x[0-9a-fA-F]{40}$/;
const TX_HASH_RE = /^0x[0-9a-fA-F]{64}$/;
// RFC 3339 date-time with the literal "Z" UTC offset, as required by the
// AgentPay canonical schemas. Fractional seconds are optional.
const RFC3339_Z_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{1,9})?Z$/;
const UUID_RE = /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/;
const INTEGER_STRING_RE = /^(0|-?[1-9][0-9]*)$/;
const BASE64_RE = /^[A-Za-z0-9+/]*={0,2}$/;

/**
 * Walk `value` against the schema and throw `CanonicalJsonError` on the
 * first violation. Returns the same value untouched on success so callers
 * can chain `encode(validate(v, S))`.
 */
export function validate<T = unknown>(value: unknown, schema: ObjectSchema): T {
  validateObject(value, schema, '');
  return value as T;
}

function validateField(value: unknown, fieldSchema: FieldSchema, path: string): void {
  if (value === null) {
    // Null acceptance is handled by the caller (validateObject), which
    // has access to the surrounding object for the nullable predicate.
    // Reaching this branch with `null` means the caller permitted it.
    return;
  }
  validateKind(value, fieldSchema.kind, path);
}

function validateKind(value: unknown, kind: FieldKind, path: string): void {
  if (typeof kind === 'string') {
    validateScalar(value, kind, path);
    return;
  }
  if ('enum' in kind) {
    if (typeof value !== 'string') {
      throw new CanonicalJsonError(path, `expected string from enum, got ${typeName(value)}`);
    }
    if (!kind.enum.includes(value)) {
      throw new CanonicalJsonError(
        path,
        `value ${JSON.stringify(value)} is not in enum {${kind.enum.join(', ')}}`,
      );
    }
    return;
  }
  if ('object' in kind) {
    validateObject(value, kind.object, path);
    return;
  }
  if ('array' in kind) {
    if (!Array.isArray(value)) {
      throw new CanonicalJsonError(path, `expected array, got ${typeName(value)}`);
    }
    for (let i = 0; i < value.length; i++) {
      const item = value[i];
      if (item === null) {
        throw new CanonicalJsonError(joinPath(path, i), 'null not permitted in array element');
      }
      validateKind(item, kind.array, joinPath(path, i));
    }
    return;
  }
  if ('union' in kind) {
    const errors: string[] = [];
    for (const member of kind.union) {
      try {
        validateKind(value, member, path);
        return;
      } catch (err) {
        if (err instanceof CanonicalJsonError) {
          errors.push(err.reason);
          continue;
        }
        throw err;
      }
    }
    throw new CanonicalJsonError(
      path,
      `value did not match any union member: ${errors.join('; ')}`,
    );
  }
}

function validateScalar(value: unknown, kind: Exclude<FieldKind, object>, path: string): void {
  switch (kind) {
    case 'string':
      if (typeof value !== 'string') {
        throw new CanonicalJsonError(path, `expected string, got ${typeName(value)}`);
      }
      assertNfc(value, path);
      return;
    case 'integer-string':
      if (typeof value !== 'string') {
        throw new CanonicalJsonError(
          path,
          `expected integer-as-string, got ${typeName(value)}`,
        );
      }
      if (!INTEGER_STRING_RE.test(value)) {
        throw new CanonicalJsonError(
          path,
          'integer-as-string must be a base-10 integer without leading zeros',
        );
      }
      return;
    case 'hex':
      if (typeof value !== 'string') {
        throw new CanonicalJsonError(path, `expected hex string, got ${typeName(value)}`);
      }
      if (value.length === 0 || value.length % 2 !== 0 || !HEX_RE.test(value)) {
        throw new CanonicalJsonError(
          path,
          'hex value must be a non-empty even-length lowercase hex string',
        );
      }
      return;
    case 'address':
      if (typeof value !== 'string') {
        throw new CanonicalJsonError(path, `expected 0x-address, got ${typeName(value)}`);
      }
      if (!ADDRESS_RE.test(value)) {
        throw new CanonicalJsonError(path, 'address must match 0x[a-fA-F0-9]{40}');
      }
      return;
    case 'tx-hash':
      if (typeof value !== 'string') {
        throw new CanonicalJsonError(path, `expected 0x tx hash, got ${typeName(value)}`);
      }
      if (!TX_HASH_RE.test(value)) {
        throw new CanonicalJsonError(path, 'tx-hash must match 0x[a-fA-F0-9]{64}');
      }
      return;
    case 'rfc3339-z':
      if (typeof value !== 'string') {
        throw new CanonicalJsonError(path, `expected RFC3339-Z timestamp, got ${typeName(value)}`);
      }
      if (!RFC3339_Z_RE.test(value)) {
        throw new CanonicalJsonError(
          path,
          'timestamp must be RFC3339 UTC with literal Z suffix',
        );
      }
      return;
    case 'uuid':
      if (typeof value !== 'string') {
        throw new CanonicalJsonError(path, `expected UUID, got ${typeName(value)}`);
      }
      if (!UUID_RE.test(value)) {
        throw new CanonicalJsonError(path, 'value is not a well-formed UUID');
      }
      return;
    case 'integer':
      if (typeof value !== 'number' || !Number.isInteger(value)) {
        throw new CanonicalJsonError(path, `expected integer, got ${typeName(value)}`);
      }
      if (value > Number.MAX_SAFE_INTEGER || value < Number.MIN_SAFE_INTEGER) {
        throw new CanonicalJsonError(
          path,
          'integers outside the IEEE-754 safe range must be passed as integer-string',
        );
      }
      return;
    case 'boolean':
      if (typeof value !== 'boolean') {
        throw new CanonicalJsonError(path, `expected boolean, got ${typeName(value)}`);
      }
      return;
    case 'base64':
      if (typeof value !== 'string') {
        throw new CanonicalJsonError(path, `expected base64 string, got ${typeName(value)}`);
      }
      if (!BASE64_RE.test(value) || value.length % 4 !== 0) {
        throw new CanonicalJsonError(path, 'value is not standard base64');
      }
      return;
    default: {
      // Exhaustiveness guard.
      const _exhaustive: never = kind;
      throw new CanonicalJsonError(path, `unknown scalar kind: ${String(_exhaustive)}`);
    }
  }
}

function validateObject(value: unknown, schema: ObjectSchema, path: string): void {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new CanonicalJsonError(
      path,
      `expected object (${schema.name}), got ${typeName(value)}`,
    );
  }
  const obj = value as Record<string, unknown>;
  const closed = schema.closed !== false;

  // Reject unknown fields first so error messages point at the actual
  // offender rather than failing later on a missing required field.
  if (closed) {
    for (const key of Object.keys(obj)) {
      if (!(key in schema.fields)) {
        throw new CanonicalJsonError(
          joinPath(path, key),
          `unknown field in ${schema.name}: ${key}`,
        );
      }
    }
  }

  for (const [name, fieldSchema] of Object.entries(schema.fields)) {
    const childPath = joinPath(path, name);
    const present = name in obj;
    const childValue = obj[name];

    if (!present || childValue === undefined) {
      if (fieldSchema.optional) continue;
      throw new CanonicalJsonError(childPath, `missing required field: ${name}`);
    }

    if (childValue === null) {
      const allowed = fieldSchema.nullable?.(obj) ?? false;
      if (!allowed) {
        throw new CanonicalJsonError(childPath, 'null is not permitted at this position');
      }
      continue;
    }

    validateField(childValue, fieldSchema, childPath);
  }
}

function assertNfc(s: string, path: string): void {
  if (s.normalize('NFC') !== s) {
    throw new CanonicalJsonError(path, 'string is not in Unicode NFC');
  }
}

function typeName(v: unknown): string {
  if (v === null) return 'null';
  if (Array.isArray(v)) return 'array';
  return typeof v;
}

File: packages/canonical-json-py/src/agentpay_canonical_json/validate.py
"""
Schema validation for canonical JSON values.
"""

import re
import unicodedata
from typing import Any

from .errors import CanonicalJsonError
from .path import join_path
from .schema import (
    ArrayKind,
    EnumKind,
    FieldKind,
    FieldSchema,
    ObjectKind,
    ObjectSchema,
    UnionKind,
)

# Validation patterns
HEX_RE = re.compile(r"^[0-9a-f]+$")
ADDRESS_RE = re.compile(r"^0x[0-9a-fA-F]{40}$")
TX_HASH_RE = re.compile(r"^0x[0-9a-fA-F]{64}$")
# RFC 3339 date-time with the literal "Z" UTC offset, as required by the
# AgentPay canonical schemas. Fractional seconds are optional.
RFC3339_Z_RE = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{1,9})?Z$")
UUID_RE = re.compile(
    r"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
)
INTEGER_STRING_RE = re.compile(r"^(0|-?[1-9][0-9]*)$")
BASE64_RE = re.compile(r"^[A-Za-z0-9+/]*={0,2}$")


def validate(value: Any, schema: ObjectSchema) -> Any:
    """
    Walk `value` against the schema and raise `CanonicalJsonError` on the
    first violation. Returns the same value untouched on success so callers
    can chain `encode(validate(v, S))`.

    Args:
        value: Value to validate
        schema: Schema to validate against

    Returns:
        The input value unchanged

    Raises:
        CanonicalJsonError: On schema violation
    """
    _validate_object(value, schema, "")
    return value


def _validate_field(value: Any, field_schema: FieldSchema, path: str) -> None:
    if value is None:
        # Null acceptance is handled by the caller (_validate_object), which
        # has access to the surrounding object for the nullable predicate.
        # Reaching this branch with `None` means the caller permitted it.
        return
    _validate_kind(value, field_schema.kind, path)


def _validate_kind(value: Any, kind: FieldKind, path: str) -> None:
    if isinstance(kind, str):
        _validate_scalar(value, kind, path)
        return

    if isinstance(kind, EnumKind):
        if not isinstance(value, str):
            raise CanonicalJsonError(path, f"expected string from enum, got {_type_name(value)}")
        if value not in kind.values:
            raise CanonicalJsonError(
                path, f"value {repr(value)} is not in enum {{{', '.join(kind.values)}}}"
            )
        return

    if isinstance(kind, ObjectKind):
        _validate_object(value, kind.schema, path)
        return

    if isinstance(kind, ArrayKind):
        if not isinstance(value, list):
            raise CanonicalJsonError(path, f"expected array, got {_type_name(value)}")
        for i, item in enumerate(value):
            if item is None:
                raise CanonicalJsonError(join_path(path, i), "null not permitted in array element")
            _validate_kind(item, kind.element, join_path(path, i))
        return

    if isinstance(kind, UnionKind):
        errors: list[str] = []
        for member in kind.members:
            try:
                _validate_kind(value, member, path)
                return
            except CanonicalJsonError as err:
                errors.append(err.reason)
                continue
        raise CanonicalJsonError(path, f"value did not match any union member: {'; '.join(errors)}")


def _validate_scalar(value: Any, kind: str, path: str) -> None:
    if kind == "string":
        if not isinstance(value, str):
            raise CanonicalJsonError(path, f"expected string, got {_type_name(value)}")
        _assert_nfc(value, path)
        return

    if kind == "integer-string":
        if not isinstance(value, str):
            raise CanonicalJsonError(path, f"expected integer-as-string, got {_type_name(value)}")
        if not INTEGER_STRING_RE.match(value):
            raise CanonicalJsonError(
                path, "integer-as-string must be a base-10 integer without leading zeros"
            )
        return

    if kind == "hex":
        if not isinstance(value, str):
            raise CanonicalJsonError(path, f"expected hex string, got {_type_name(value)}")
        if not value or len(value) % 2 != 0 or not HEX_RE.match(value):
            raise CanonicalJsonError(
                path, "hex value must be a non-empty even-length lowercase hex string"
            )
        return

    if kind == "address":
        if not isinstance(value, str):
            raise CanonicalJsonError(path, f"expected 0x-address, got {_type_name(value)}")
        if not ADDRESS_RE.match(value):
            raise CanonicalJsonError(path, "address must match 0x[a-fA-F0-9]{40}")
        return

    if kind == "tx-hash":
        if not isinstance(value, str):
            raise CanonicalJsonError(path, f"expected 0x tx hash, got {_type_name(value)}")
        if not TX_HASH_RE.match(value):
            raise CanonicalJsonError(path, "tx-hash must match 0x[a-fA-F0-9]{64}")
        return

    if kind == "rfc3339-z":
        if not isinstance(value, str):
            raise CanonicalJsonError(path, f"expected RFC3339-Z timestamp, got {_type_name(value)}")
        if not RFC3339_Z_RE.match(value):
            raise CanonicalJsonError(path, "timestamp must be RFC3339 UTC with literal Z suffix")
        return

    if kind == "uuid":
        if not isinstance(value, str):
            raise CanonicalJsonError(path, f"expected UUID, got {_type_name(value)}")
        if not UUID_RE.match(value):
            raise CanonicalJsonError(path, "value is not a well-formed UUID")
        return

    if kind == "integer":
        if not isinstance(value, int) or isinstance(value, bool):
            raise CanonicalJsonError(path, f"expected integer, got {_type_name(value)}")
        # Python int is arbitrary precision, but we follow the JavaScript safe range
        MAX_SAFE = 2**53 - 1
        MIN_SAFE = -(2**53 - 1)
        if value > MAX_SAFE or value < MIN_SAFE:
            raise CanonicalJsonError(
                path, "integers outside the IEEE-754 safe range must be passed as integer-string"
            )
        return

    if kind == "boolean":
        if not isinstance(value, bool):
            raise CanonicalJsonError(path, f"expected boolean, got {_type_name(value)}")
        return

    if kind == "base64":
        if not isinstance(value, str):
            raise CanonicalJsonError(path, f"expected base64 string, got {_type_name(value)}")
        if not BASE64_RE.match(value) or len(value) % 4 != 0:
            raise CanonicalJsonError(path, "value is not standard base64")
        return

    raise CanonicalJsonError(path, f"unknown scalar kind: {kind}")


def _validate_object(value: Any, schema: ObjectSchema, path: str) -> None:
    if not isinstance(value, dict) or value is None:
        raise CanonicalJsonError(path, f"expected object ({schema.name}), got {_type_name(value)}")

    closed = schema.closed

    # Reject unknown fields first so error messages point at the actual
    # offender rather than failing later on a missing required field.
    if closed:
        for key in value.keys():
            if key not in schema.fields:
                raise CanonicalJsonError(
                    join_path(path, key), f"unknown field in {schema.name}: {key}"
                )

    for name, field_schema in schema.fields.items():
        child_path = join_path(path, name)
        present = name in value
        child_value = value.get(name)

        if not present:
            if field_schema.optional:
                continue
            raise CanonicalJsonError(child_path, f"missing required field: {name}")

        if child_value is None:
            allowed = field_schema.nullable(value)
            if not allowed:
                raise CanonicalJsonError(child_path, "null is not permitted at this position")
            continue

        _validate_field(child_value, field_schema, child_path)


def _assert_nfc(s: str, path: str) -> None:
    if unicodedata.normalize("NFC", s) != s:
        raise CanonicalJsonError(path, "string is not in Unicode NFC")


def _type_name(v: Any) -> str:
    if v is None:
        return "null"
    if isinstance(v, list):
        return "array"
    if isinstance(v, bool):
        return "boolean"
    if isinstance(v, int):
        return "integer"
    if isinstance(v, str):
        return "string"
    if isinstance(v, dict):
        return "object"
    return type(v).__name__
